File: valid_parentheses/valid-parentheses.py
"""Valid Parenthesis

Time Complexity: O(n)
Space Complexity: O(n)
"""

class Solution:
    def is_balanced(self, string):
        open = tuple('[({')
        close = tuple('])}')
        brackets = dict(zip(open, close))
        stack = []

        for char in string:
            if char in open:
                stack.append(char)
            elif not stack and char in close:
                return False     
            elif stack:
                peek = stack[-1]
                if char in close and char == brackets[peek]:
                    stack.pop()
        return not stack
    
def main():
    s = Solution()
    string = input('Enter string: ')
    print(s.is_balanced(string))

if __name__ == "__main__":
    main()

File: valid_parentheses/valid-parentheses.cpp
/* Valid Parentheses
*
*  Time Complexity: O(n)
* Space Complexity: O(n)
*/

#include <iostream>
#include <stack>
#include <string>

class Solution {
    public: 
        bool is_pair(char open, char close) {
            if (open == '[' && close == ']') return true;
            if (open == '(' && close == ')') return true;
            if (open == '{' && close == '}') return true;
            else return false;
        }

        bool is_balanced(std::string str) {
            std::stack<char> stack;

            for (int i = 0; i < str.length(); i++) {
                if (str[i] == '[' || str[i] == '(' || str[i] == '{') {
                    stack.push(str[i]);
                }
                else if (str[i] == ']' || str[i] == ')' || str[i] == '}') {
                    if (stack.empty()) return false;
                    else if (is_pair(stack.top(), str[i])) stack.pop();
                    else return false;
                }
            }
            return stack.empty();
        }
};

int main() {
    Solution s;
    std::string str;
    
    std::cout << "Enter string: ";
    std::cin >> str;
    std::cout << std::boolalpha << s.is_balanced(str) << '\n';
}